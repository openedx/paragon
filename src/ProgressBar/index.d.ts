import React from 'react';
import { ProgressBarProps as BaseProgressBarProps } from 'react-bootstrap/ProgressBar';

export interface ProgressBarProps extends BaseProgressBarProps {}

export interface ProgressBarAnnotatedProps extends ProgressBarProps {
    threshold?: number;
    thresholdLabel?: React.ReactNode;
    thresholdVariant?: 'dark' | 'warning' | 'success' | 'error';
    progressHint?: React.ReactNode;
    thresholdHint?: React.ReactNode;
}

declare const ProgressBar: React.FC<ProgressBarProps> & {
    Annotated: React.FC<ProgressBarAnnotatedProps>;
};

export default ProgressBar;
