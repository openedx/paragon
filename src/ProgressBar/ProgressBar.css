/**
 * Do not edit directly
 * Generated on Wed, 01 Mar 2023 00:27:20 GMT
 */

:root {
  --pgn-elevation-progress-bar-box-shadow: none;
  --pgn-transition-progress-bar-bar-transition: width .6s ease;
  --pgn-transition-progress-bar-bar-animation-timing: 1s linear infinite;
  --pgn-spacing-progress-bar-hint-annotation-gap: .5rem;
  --pgn-size-progress-bar-threshold-circle: .5625rem;
  --pgn-size-progress-bar-border-radius: 0;
  --pgn-size-progress-bar-border-width: 1px;
  --pgn-size-progress-bar-height-annotated: .3125rem;
  --pgn-size-progress-bar-height-base: 1rem;
  --pgn-typography-progress-bar-font-size: calc(var(--pgn-typography-font-size-base) * .75);
}
