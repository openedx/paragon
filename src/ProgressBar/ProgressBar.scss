@import "variables";
@import "~bootstrap/scss/progress";

.progress {
  border: $progress-bar-border-width solid $progress-bar-border-color;
}

.progress-bar {
  background-color: $progress-bar-bg;
}

.pgn__progress-annotated {
  width: 100%;
  position: relative;
  overflow: visible;

  .progress {
    overflow: visible;
    background-color: $light-300;
    height: $annotated-progress-height;
    border: none;

    .progress-bar {
      background-color: $annotated-progress-bar-bg;
      overflow: visible;
      position: relative;
    }

    .pgn__progress-tick--white::after,
    .pgn__progress-tick--black::after {
      content: "";
      position: absolute;
      height: $annotated-progress-height;
      width: 1px;
      right: 0;
    }

    .pgn__progress-tick--white::after {
      background: $light-300;
    }

    .pgn__progress-tick--black::after {
      background: $primary-500;
    }

    @each $name, $color in $progress-colors {
      .pgn__progress-bar--#{$name} {
        background-color: $color;

        &::after {
          content: "";
          display: block;
          background: $color;
          position: absolute;
          top: -(calc($progress-threshold-circle / 2) - calc($annotated-progress-height / 2));
          right: -(calc($progress-threshold-circle / 2));
          width: $progress-threshold-circle;
          height: $progress-threshold-circle;
          border-radius: calc($progress-threshold-circle / 2);
          z-index: 1;
        }
      }
    }
  }

  .progress::before,
  .progress::after {
    position: absolute;
    content: "";
    height: $annotated-progress-height;
    width: 1px;
    background: $primary-500;
  }

  .progress::after {
    right: 0;
  }

  .pgn__progress-info {
    display: inline-block;
    position: relative;
  }

  .pgn__progress-hint {
    box-sizing: border-box;
    padding: 0 $progress-hint-annotation-gap;
    font-size: $small-font-size;
  }
}
