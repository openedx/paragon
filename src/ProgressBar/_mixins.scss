@mixin progress-bar-variant($bg) {
  background-color: $bg;

  &::after {
    content: "";
    display: block;
    background: $bg;
    position: absolute;
    // stylelint-disable-next-line max-line-length
    top: calc(var(--pgn-size-progress-bar-threshold-circle) / 2 - var(--pgn-size-progress-bar-height-annotated) / 2);
    right: calc(var(--pgn-size-progress-bar-threshold-circle) / -2);
    width: var(--pgn-size-progress-bar-threshold-circle);
    height: var(--pgn-size-progress-bar-threshold-circle);
    border-radius: calc(var(--pgn-size-progress-bar-threshold-circle) / 2);
    z-index: 1;
  }
}
