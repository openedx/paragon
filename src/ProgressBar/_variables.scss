// Progress bars

$progress-height:                   1rem !default;
$annotated-progress-height:         .3125rem !default;
$progress-font-size:                $font-size-base * .75 !default;
$progress-bg:                       transparent !default;
$progress-border-radius:            0 !default;
$progress-box-shadow:               none !default;
$progress-bar-color:                $white !default;
$progress-bar-bg:                   theme-color("accent-a") !default;
$annotated-progress-bar-bg:         theme-color("dark") !default;
$progress-bar-animation-timing:     1s linear infinite !default;
$progress-bar-transition:           width .6s ease !default;
$progress-bar-border-width:         1px !default;
$progress-bar-border-color:         $gray-500 !default;
$progress-threshold-circle:         .5625rem !default;
$progress-hint-annotation-gap:      .5rem !default;
$progress-colors: (
  "dark":    $primary-500,
  "success": $success-500,
  "error":   $danger-500,
  "warning": $accent-b,
) !default;
