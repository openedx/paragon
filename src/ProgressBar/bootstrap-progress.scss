@if $enable-transitions {
  @keyframes progress-bar-stripes {
    from { background-position: var(--pgn-size-progress-bar-height-base) 0; }
    to { background-position: 0 0; }
  }
}

.progress {
  display: flex;
  height: var(--pgn-size-progress-bar-height-base);
  overflow: hidden;
  line-height: 0;
  background-color: var(--pgn-color-progress-bar-bg);
  box-shadow: var(--pgn-elevation-progress-bar-box-shadow);

  @include font-size(var(--pgn-typography-progress-bar-font-size));
  @include border-radius(var(--pgn-size-progress-bar-border-radius));
}

.progress-bar {
  display: flex;
  flex-direction: column;
  justify-content: center;
  overflow: hidden;
  color: var(--pgn-color-progress-bar-bar-base);
  text-align: center;
  white-space: nowrap;
  background-color: var(--pgn-color-progress-bar-bar-bg-base);

  @include transition(var(--pgn-transition-progress-bar-bar-transition));
}

.progress-bar-striped {
  @include gradient-striped();

  background-size: var(--pgn-size-progress-bar-height-base) var(--pgn-size-progress-bar-height-base);
}

@if $enable-transitions {
  .progress-bar-animated {
    animation: var(--pgn-transition-progress-bar-bar-animation-timing) progress-bar-stripes;

    @if $enable-prefers-reduced-motion-media-query {
      @media (prefers-reduced-motion: reduce) {
        animation: none;
      }
    }
  }
}
