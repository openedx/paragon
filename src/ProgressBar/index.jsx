import React, { useCallback, useEffect } from 'react';
import ProgressBarBase from 'react-bootstrap/ProgressBar';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Annotation from '../Annotation';
import { getOffsetStyles, placeInfoAtZero } from './utils';

export const ANNOTATION_CLASS = 'pgn__annotation';
const HINT_SWAP_PERCENT = 50;
const PROGRESS_DEFAULT_VARIANT = 'warning';
const THRESHOLD_DEFAULT_VARIANT = 'dark';
const VARIANTS = [
  'dark',
  'warning',
  'success',
  'error',
];

function ProgressBar(props) {
  return <ProgressBarBase {...props} />;
}

function ProgressBarAnnotated({
  now,
  label,
  variant,
  threshold,
  thresholdLabel,
  thresholdVariant,
  progressHint,
  thresholdHint,
  ...props
}) {
  const progressInfoRef = React.useRef();
  const thresholdInfoRef = React.useRef();
  const thresholdPercent = (threshold || 0) - (now || 0);
  const isProgressHintAfter = now < HINT_SWAP_PERCENT;
  const isThresholdHintAfter = threshold < HINT_SWAP_PERCENT;
  const progressColor = VARIANTS.includes(variant) ? variant : PROGRESS_DEFAULT_VARIANT;
  const thresholdColor = VARIANTS.includes(thresholdVariant) ? thresholdVariant : THRESHOLD_DEFAULT_VARIANT;
  const direction = window.getComputedStyle(document.body).getPropertyValue('direction');

  const positionAnnotations = useCallback(() => {
    placeInfoAtZero(progressInfoRef, direction, isProgressHintAfter, ANNOTATION_CLASS);
    placeInfoAtZero(thresholdInfoRef, direction, isThresholdHintAfter, ANNOTATION_CLASS);
  }, [direction, isProgressHintAfter, isThresholdHintAfter]);

  useEffect(() => {
    positionAnnotations();
    const observer = new ResizeObserver(() => {
      positionAnnotations();
    });
    const progressInfoEl = progressInfoRef.current;
    observer.observe(progressInfoEl);
    return () => progressInfoEl && observer.unobserve(progressInfoEl);
  }, [positionAnnotations]);

  const getHint = (text) => (
    <span className="pgn__progress-hint" data-testid="progress-hint">
      {text}
    </span>
  );

  return (
    <div className="pgn__progress-annotated">
      {!!label && (
        <div
          className="pgn__progress-info"
          style={getOffsetStyles(now, direction)}
          ref={progressInfoRef}
        >
          {!isProgressHintAfter && getHint(progressHint)}
          <Annotation variant={progressColor}>
            {label}
          </Annotation>
          {isProgressHintAfter && getHint(progressHint)}
        </div>
      )}
      <ProgressBarBase>
        <ProgressBarBase
          {...props}
          now={now}
          className={classNames(
            `pgn__progress-bar--${progressColor}`,
            thresholdPercent > 0 ? 'pgn__progress-tick--white' : 'pgn__progress-tick--black',
          )}
          srOnly
        />
        {!!threshold && (
          <ProgressBarBase
            now={thresholdPercent}
            className={`pgn__progress-bar--${thresholdColor}`}
            srOnly
          />
        )}
      </ProgressBarBase>
      {(!!threshold && !!thresholdLabel) && (
        <div
          className="pgn__progress-info"
          style={getOffsetStyles(threshold, direction)}
          ref={thresholdInfoRef}
        >
          {!isThresholdHintAfter && getHint(thresholdHint)}
          <Annotation
            arrowPlacement="top"
            variant={thresholdColor}
          >
            {thresholdLabel}
          </Annotation>
          {isThresholdHintAfter && getHint(thresholdHint)}
        </div>
      )}
    </div>
  );
}

ProgressBarAnnotated.propTypes = {
  /** Current value of progress. */
  now: PropTypes.number,
  /** Show label that represents visual percentage. */
  label: PropTypes.node,
  /** The `ProgressBar` style variant to use. */
  variant: PropTypes.oneOf(VARIANTS),
  /** Specifies an additional `className` to add to the base element. */
  className: PropTypes.string,
  /** Threshold current value. */
  threshold: PropTypes.number,
  /** Specifies label for `threshold`. */
  thresholdLabel: PropTypes.node,
  /** Variant for threshold value. */
  thresholdVariant: PropTypes.oneOf(VARIANTS),
  /** Text near the progress annotation. */
  progressHint: PropTypes.node,
  /** Text near the threshold annotation. */
  thresholdHint: PropTypes.node,
};

ProgressBarAnnotated.defaultProps = {
  now: undefined,
  label: undefined,
  variant: PROGRESS_DEFAULT_VARIANT,
  className: undefined,
  threshold: undefined,
  thresholdLabel: undefined,
  thresholdVariant: THRESHOLD_DEFAULT_VARIANT,
  progressHint: undefined,
  thresholdHint: undefined,
};

ProgressBar.Annotated = ProgressBarAnnotated;
export default ProgressBar;
