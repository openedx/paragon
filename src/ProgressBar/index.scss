@import "variables";
@import "~bootstrap/scss/progress";

.progress {
  border: $progress-bar-border-width solid $progress-bar-border-color;
}

.progress-bar {
  background-color: $progress-bar-bg;
}

.pgn__progress-annotated {
  width: 100%;
  position: relative;
  overflow: visible;
  padding: 3.125rem 0;

  .progress {
    overflow: visible;
    background-color: $light-300;
    height: $annotated-progress-height;
    border: none;

    .progress-bar {
      background-color: $annotated-progress-bar-bg;
      overflow: visible;
      position: relative;
    }

    .pgn__progress-tick--white::after,
    .pgn__progress-tick--black::after {
      content: "";
      position: absolute;
      height: $annotated-progress-height;
      width: 1px;
      right: 0;
    }

    .pgn__progress-tick--white::after {
      background: $light-300;
    }

    .pgn__progress-tick--black::after {
      background: $primary-500;
    }

    @each $name, $color in $progress-colors {
      .pgn__progress-bar--#{$name} {
        background-color: $color;

        &::after {
          content: "";
          display: block;
          background: $color;
          position: absolute;
          top: -(calc($progress-threshold-circle / 2) - calc($annotated-progress-height / 2));
          right: -(calc($progress-threshold-circle / 2));
          width: $progress-threshold-circle;
          height: $progress-threshold-circle;
          border-radius: calc($progress-threshold-circle / 2);
          z-index: 1;

          [dir="rtl"] & {
            left: -(calc($progress-threshold-circle / 2));
            right: auto;
          }
        }
      }
    }
  }

  .progress::before,
  .progress::after {
    position: absolute;
    content: "";
    height: $annotated-progress-height;
    width: 1px;
    background: $primary-500;
  }

  .progress::after {
    right: 0;
  }

  .pgn__progress-info {
    position: absolute;
    display: flex;
    align-items: baseline;

    &:first-child {
      top: 0;
    }
  }

  .pgn__progress-hint {
    box-sizing: border-box;
    padding: 0 $progress-hint-annotation-gap;
    font-size: $small-font-size;
  }
}
