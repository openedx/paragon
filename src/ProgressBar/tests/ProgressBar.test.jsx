import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import ProgressBar, { ANNOTATION_CLASS } from '..';

const ref = {
  current: {
    style: {
      marginLeft: '',
    },
    children: [
      {
        className: `${ANNOTATION_CLASS} someClass`,
        getBoundingClientRect: () => ({
          width: 24.58123,
        }),
      },
      {
        className: 'anotherClass',
        getBoundingClientRect: () => ({
          width: 55.96844,
        }),
      },
    ],
  },
};

function ProgressBarElement(props) {
  return (
    <ProgressBar.Annotated
      now={20}
      label="20%"
      threshold={50}
      thresholdLabel="50%"
      {...props}
    />
  );
}

describe('<ProgressBar.Annotated />', () => {
  describe('correct rendering', () => {
    it('renders without props', () => {
      const tree = renderer.create((
        <ProgressBar.Annotated />
      )).toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('renders info blocks with calculated margins', () => {
      const useReferenceSpy = jest.spyOn(React, 'useRef').mockReturnValue(ref);
      mount(<ProgressBarElement />);
      expect(useReferenceSpy).toHaveBeenCalledTimes(2);
      expect(ref.current.style.marginLeft).not.toBeFalsy();
    });
    it('renders correct variant for progress bar and annotation', () => {
      const progressVariant = 'success';
      const thresholdVariant = 'error';
      const wrapper = mount(<ProgressBarElement variant={progressVariant} thresholdVariant={thresholdVariant} />);
      expect(wrapper.find(`.pgn__annotation-${progressVariant}-bottom`).length).toBeGreaterThan(0);
      expect(wrapper.find(`.pgn__progress-bar--${progressVariant}`).length).toBeGreaterThan(0);
      expect(wrapper.find(`.pgn__annotation-${thresholdVariant}-top`).length).toBeGreaterThan(0);
      expect(wrapper.find(`.pgn__progress-bar--${thresholdVariant}`).length).toBeGreaterThan(0);
    });
    it('renders default variant for progress bar and annotation', () => {
      const wrapper = mount(<ProgressBarElement />);
      expect(wrapper.find('.pgn__annotation-warning-bottom').length).toBeGreaterThan(0);
      expect(wrapper.find('.pgn__progress-bar--warning').length).toBeGreaterThan(0);
      expect(wrapper.find('.pgn__annotation-dark-top').length).toBeGreaterThan(0);
      expect(wrapper.find('.pgn__progress-bar--dark').length).toBeGreaterThan(0);
    });
    it('renders hints with text', () => {
      const progressHint = 'first hint';
      const thresholdHint = 'second hint';
      const wrapper = mount(<ProgressBarElement progressHint={progressHint} thresholdHint={thresholdHint} />);
      expect(wrapper.find('.pgn__progress-hint').length).toBeGreaterThan(0);
      expect(wrapper.find('.pgn__progress-hint').at(0).text()).toEqual(progressHint);
      expect(wrapper.find('.pgn__progress-hint').at(1).text()).toEqual(thresholdHint);
    });
    it('renders hints with text', () => {
      const wrapper = mount(<ProgressBarElement now={30} threshold={40} />);
      expect(wrapper.find('.pgn__progress-hint').length).toBeGreaterThan(0);
      expect(wrapper.find('.pgn__progress-info').get(0).props.children[0]).toEqual(false);
      expect(wrapper.find('.pgn__progress-info').get(1).props.children[0]).toEqual(false);
      wrapper.setProps({ now: 70, threshold: 70 });
      expect(wrapper.find('.pgn__progress-info').get(0).props.children[2]).toEqual(false);
      expect(wrapper.find('.pgn__progress-info').get(1).props.children[2]).toEqual(false);
    });
  });
});
