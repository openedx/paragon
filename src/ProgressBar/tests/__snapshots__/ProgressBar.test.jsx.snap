// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProgressBar.Annotated /> correct rendering renders without props 1`] = `
<div
  className="pgn__progress-annotated"
>
  <div
    className="progress"
  >
    <div
      aria-valuemax={100}
      aria-valuemin={0}
      className="pgn__progress-bar--warning pgn__progress-tick--black progress-bar"
      role="progressbar"
      style={
        Object {
          "width": "NaN%",
        }
      }
    >
      <span
        className="sr-only"
      />
    </div>
  </div>
</div>
`;
